### ISSUE TITLE
ec2 module: profile option ignored

### ISSUE THREAD

**Issue by maxrothman** [ansible/ansible#19112]
_From @maxrothman on July 9, 2015 20:22_

**Issue Type**: Bug Report
**Component Name**: ec2 module
**Ansible Version**: 1.9
**Ansible Configuration**: default
**Environment**: OSX 10.9.5
**Summary**: The "profile" option in the ec2 module is ignored

**Steps To Reproduce**:
Create a playbook like so and run it. The account in which `subnet-0000000` exists must not be in your default aws profile (defined in `~/.aws/{credentials,config}`:

```
- name: Make an instance
  hosts: all
  connection: local
  gather_facts: False
  tasks:
    - name: Manage instances
      ec2:
        profile: myprofile
        region: us-east-1
        wait: yes
        group_id: sg-000000
        key_name: mykey
        vpc_subnet_id: subnet-0000000
        instance_type: m3.medium
        instance_tags:
          Name: my-test
        image: ami-0000000
        volumes: 
          - device_name: /dev/sdb
            ephemeral: ephemeral0
```

**Expected Results**: An instance is created as specified

**Actual Results**:
The following error occurs:

```
failed: [localhost] => {"failed": true, "parsed": false}
Traceback (most recent call last):
  File "/Users/me/.ansible/tmp/ansible-tmp-1436472344.45-200995340429191/ec2", line 3054, in <module>
    main()
  File "/Users/me/.ansible/tmp/ansible-tmp-1436472344.45-200995340429191/ec2", line 1246, in main
    (instance_dict_array, new_instance_ids, changed) = create_instances(module, ec2, vpc)
  File "/Users/me/.ansible/tmp/ansible-tmp-1436472344.45-200995340429191/ec2", line 789, in create_instances
    vpc_id = vpc.get_all_subnets(subnet_ids=[vpc_subnet_id])[0].vpc_id
  File "/Library/Python/2.7/site-packages/boto/vpc/__init__.py", line 1153, in get_all_subnets
    return self.get_list('DescribeSubnets', params, [('item', Subnet)])
  File "/Library/Python/2.7/site-packages/boto/connection.py", line 1185, in get_list
    raise self.ResponseError(response.status, response.reason, body)
boto.exception.EC2ResponseError: EC2ResponseError: 400 Bad Request
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>InvalidSubnetID.NotFound</Code><Message>The subnet ID 'subnet-000000' does not exist</Message></Error></Errors><RequestID>xxxxxxxxxxxxxxxxxx</RequestID></Response>
```

Changing the profile in which that subnet exists to the default profile allows the playbook to run as expected.

_Copied from original issue: ansible/ansible-modules-core#1701_

**Comment by maxrothman** [comment]
Any progress on this? This is a major issue for anyone using this module with multiple profiles.

**Comment by maxrothman** [comment]
_From @mabroor on October 1, 2015 15:38_

+1 for this.

**Comment by maxrothman** [comment]
_From @shuji on December 7, 2015 8:25_

me too

**Comment by maxrothman** [comment]
_From @vegaed on December 7, 2015 13:12_

+1

**Comment by maxrothman** [comment]
_From @sqlcook on December 8, 2015 6:33_

it looks to be profile , and security_token are all ignored.
looking at the source code i cannot find reference to it either.

i do see key and secret being pulled here:
https://github.com/ansible/ansible-modules-core/blob/stable-1.9/cloud/amazon/ec2.py#L1211

**Comment by maxrothman** [comment]
_From @nelsonenzo on June 2, 2016 21:21_

+1 two hours of my life I wish I had back :/

**Comment by maxrothman** [comment]
_From @ansibot on August 1, 2016 3:13_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by maxrothman** [comment]
_From @chevectra87 on August 3, 2016 22:29_

+1

**Comment by maxrothman** [comment]
_From @ikrauchanka on August 12, 2016 15:0_

+1

**Comment by maxrothman** [comment]
_From @twistedpair on August 28, 2016 15:36_

This has been broken for over **13 months**! If it's not going to be fixed, at least remove any mention of it from the [docs](https://docs.ansible.com/ansible/ec2_module.html) so people that read the manual don't waste their time.

**Comment by maxrothman** [comment]
_From @ansibot on August 28, 2016 15:44_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by maxrothman** [comment]
_From @courtneywhalen on August 31, 2016 22:55_

+1

**Comment by maxrothman** [comment]
_From @ansibot on September 12, 2016 21:23_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by maxrothman** [comment]
_From @ansibot on October 6, 2016 15:56_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by maxrothman** [comment]
_From @ansibot on October 22, 2016 9:24_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by maxrothman** [comment]
_From @ansibot on November 6, 2016 10:10_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by maxrothman** [comment]
_From @ansibot on November 22, 2016 10:18_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by maxrothman** [comment]
This issue has been open for over a year. Is it scheduled to be worked on anytime soon?

**Comment by maxrothman** [comment]
_From @ansibot on December 9, 2016 19:46_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by ansibot** [comment]
@maxrothman Greetings! Thanks for taking the time to open this issue. In order for the community to handle your issue effectively, we need a bit more information. 

Here are the items we could not find in your description:
- component name

Please set the description of this issue with this template:
https://raw.githubusercontent.com/ansible/ansible/devel/.github/ISSUE_TEMPLATE.md

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by marcelocarlos** [comment]
This is a show stopper for us, any plan to fix it?

The "component name" is in the description and it is "ec2_module", but it also affects any aws-related module.

Can you please fix/update the tags and prioritise this bugfix as it's being around for almost two years now?

It is still an issue in version 2.3.0:

```
$ ansible --version
ansible 2.3.0.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 18 2016, 07:03:39) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.42.1)]
```

**Comment by ansibot** [comment]
cc @lwade @tgerla
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by s-hertel** [comment]
@marcelocarlos Hi. I believe the tags are correct. I'm struggling to reproduce this on 2.3, can you provide a reproducer and the full traceback/error? I'm able to successfully override my default profile and environment variable credentials using the `profile:` module argument. Are you using multiple accounts?

**Comment by s-hertel** [comment]
@maxrothman Can you test this on ansible version 2.2 or 2.3? A lot has changed and I cannot reproduce this. What version of boto are you using?

needs_info

**Comment by ghost** [comment]
@s-hertel it works fine if you use profiles stored in `.aws/credentials`. However, if you use a profile defined in `.aws/config` it doesn't work.

For example (in `.aws/config`):

```
[profile staging]
role_arn = arn:aws:iam::1234567890:role/my-role
source_profile = my-profile
```

Using `profile: staging` in ansible won't work.

**Comment by marcelocarlos** [comment]
Profiles in `.aws/config` don't work as mentioned above. Tested again:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: boto.provider.ProfileNotFoundError: Profile "my-other-profile" not found!
```

Where `my-other-profile` is in `.aws/config`:

```
[profile my-other-profile]
role_arn = arn:aws:iam::1234567890:role/my-role
source_profile = my-profile
```

If I use `my-profile` which is in `.aws/credentials`, it works just fine.

Ansible version:

```
$ ansible --version
ansible 2.3.1.0
  config file =
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jul 18 2017, 09:17:00) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)
```

**Comment by s-hertel** [comment]
@marcelocarlosbr @marcelocarlos Thanks, that's helpful! This is because boto3 style credentials are being used for boto(2) modules. If you use a module that only uses boto3 this should work (just tested that to verify). Things are actively being ported to boto3 which will eventually eliminate this problem.

**Comment by ansibot** [comment]
cc @willthames
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibotdev** [comment]
@maxrothman You have not responded to information requests in this issue so we will assume it no longer affects you. If you are still interested in this, please create a new issue with the requested information.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by wutzkachel** [comment]
I have the same problem.

**Comment by wutzkachel** [comment]
This is the playbook:
------------------------
- name: ProvisionEc2CentOs7Instance
  hosts: localhost
  connection: local
  gather_facts: false
  vars_files:
    - awscredentials.yml
  tasks:
    - name: Launch instance
      ec2:
        profile: AWSCockpitMonitoringDevelEnv
        region: "{{ region }}"
        key_name: "{{ key_pair_name }}"
        instance_type: "{{ ec2_instance_type }}"
        image: "{{image_ID}}"
        vpc_subnet_id: "{{ subnet_ID }}"
        assign_public_ip: no
        group: "{{ security_group_name }}"
        user_data: "{{lookup('file', 'cloudconfig.txt')}}"
        wait: true
        count_tag:
          Name: testrechner
        exact_count: 1
        instance_tags:
          Creator: michaelmckirchner
          Name: testrechner
          Function: testrechner
        volumes:
          - device_name: /dev/sda1
            delete_on_termination: true
            volume_type: gp2
            volume_size: 10
          - device_name: /dev/sdf
            delete_on_termination: true
            encrypted: true
            kms_key_id: "{{ ec2volume_encryption_key }}"
            volume_type: gp2
            volume_size: 20
            async: 7200
            poll: 10
      register: ec2info

**Comment by wutzkachel** [comment]
cat ~/.aws/credentials
...
[AWSCockpitMonitoringDevelEnv]
output = json
region = eu-central-1
aws_access_key_id = TheAccessKey
aws_secret_access_key = TheSecretKey
aws_session_token = TheSessionToken
expiration = 2019-01-25 12:59:27+00:00
....

**Comment by wutzkachel** [comment]
he full traceback is:
Traceback (most recent call last):
  File "/home/michael/.ansible/tmp/ansible-tmp-1548417736.96-186440713350129/AnsiballZ_ec2.py", line 113, in <module>
    _ansiballz_main()
  File "/home/michael/.ansible/tmp/ansible-tmp-1548417736.96-186440713350129/AnsiballZ_ec2.py", line 105, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/michael/.ansible/tmp/ansible-tmp-1548417736.96-186440713350129/AnsiballZ_ec2.py", line 48, in invoke_module
    imp.load_module('__main__', mod, module, MOD_DESC)
  File "/tmp/ansible_ec2_payload_PV2mTT/__main__.py", line 1692, in <module>
  File "/tmp/ansible_ec2_payload_PV2mTT/__main__.py", line 1678, in main
  File "/tmp/ansible_ec2_payload_PV2mTT/__main__.py", line 887, in enforce_count
  File "/tmp/ansible_ec2_payload_PV2mTT/__main__.py", line 568, in find_running_instances_by_count_tag
  File "/tmp/ansible_ec2_payload_PV2mTT/__main__.py", line 600, in get_reservations
  File "/home/michael/.local/share/virtualenvs/PipenvAnsible27Python27-TIMLuz9F/lib/python2.7/site-packages/boto/vpc/__init__.py", line 1152, in get_all_subnets
    return self.get_list('DescribeSubnets', params, [('item', Subnet)])
  File "/home/michael/.local/share/virtualenvs/PipenvAnsible27Python27-TIMLuz9F/lib/python2.7/site-packages/boto/connection.py", line 1186, in get_list
    raise self.ResponseError(response.status, response.reason, body)
boto.exception.EC2ResponseError: EC2ResponseError: 401 Unauthorized
<?xml version="1.0" encoding="UTF-8"?>
<Response><Errors><Error><Code>AuthFailure</Code><Message>AWS was not able to validate the provided access credentials</Message></Error></Errors><RequestID>5ba56ef0-ff79-4587-a5b2-e74a2b478bf8</RequestID></Response>

fatal: [localhost]: FAILED! => {
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/home/michael/.ansible/tmp/ansible-tmp-1548417736.96-186440713350129/AnsiballZ_ec2.py\", line 113, in <module>\n    _ansiballz_main()\n  File \"/home/michael/.ansible/tmp/ansible-tmp-1548417736.96-186440713350129/AnsiballZ_ec2.py\", line 105, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/michael/.ansible/tmp/ansible-tmp-1548417736.96-186440713350129/AnsiballZ_ec2.py\", line 48, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_ec2_payload_PV2mTT/__main__.py\", line 1692, in <module>\n  File \"/tmp/ansible_ec2_payload_PV2mTT/__main__.py\", line 1678, in main\n  File \"/tmp/ansible_ec2_payload_PV2mTT/__main__.py\", line 887, in enforce_count\n  File \"/tmp/ansible_ec2_payload_PV2mTT/__main__.py\", line 568, in find_running_instances_by_count_tag\n  File \"/tmp/ansible_ec2_payload_PV2mTT/__main__.py\", line 600, in get_reservations\n  File \"/home/michael/.local/share/virtualenvs/PipenvAnsible27Python27-TIMLuz9F/lib/python2.7/site-packages/boto/vpc/__init__.py\", line 1152, in get_all_subnets\n    return self.get_list('DescribeSubnets', params, [('item', Subnet)])\n  File \"/home/michael/.local/share/virtualenvs/PipenvAnsible27Python27-TIMLuz9F/lib/python2.7/site-packages/boto/connection.py\", line 1186, in get_list\n    raise self.ResponseError(response.status, response.reason, body)\nboto.exception.EC2ResponseError: EC2ResponseError: 401 Unauthorized\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Response><Errors><Error><Code>AuthFailure</Code><Message>**AWS was not able to validate the provided access credentials**</Message></Error></Errors><RequestID>5ba56ef0-ff79-4587-a5b2-e74a2b478bf8</RequestID></Response>\n", 
    "module_stdout": "reply: 'HTTP/1.1 200 Connection established\\r\\n'\nheader: Connection: close\r\nsend: u'GET https://ec2.eu-central-1.amazonaws.com/?Action=DescribeSubnets&SubnetId.1=subnet-03dfb6889a3e2ef29&Version=2014-10-01 HTTP/1.1\\r\\nAccept-Encoding: identity\\r\\nContent-Length: 0\\r\\nHost: ec2.eu-central-1.amazonaws.com\\r\\nAuthorization: AWS4-HMAC-SHA256 Credential=SomeInformation/20190125/eu-central-1/ec2/aws4_request,SignedHeaders=host;x-amz-date,Signature=SomeInformation\\r\\nX-Amz-Date: 20190125T120219Z\\r\\nUser-Agent: Boto/2.49.0 Python/2.7.15 Linux/\\r\\n\\r\\n'\nreply: 'HTTP/1.1 401 Unauthorized\\r\\n'\nheader: Transfer-Encoding: chunked\r\nheader: Date: Fri, 25 Jan 2019 12:02:18 GMT\r\nheader: Server: AmazonEC2\r\n", 
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", 
    "rc": 1
}

**Comment by wutzkachel** [comment]
ansible 2.7.6
  config file = /home/michael/DockerSwarmMonitoringCluster/ansible.cfg
  configured module search path = [u'/home/michael/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/michael/.local/share/virtualenvs/PipenvAnsible27Python27-TIMLuz9F/lib/python2.7/site-packages/ansible
  executable location = /home/michael/.local/share/virtualenvs/PipenvAnsible27Python27-TIMLuz9F/bin/ansible
  python version = 2.7.15 (default, Jan 10 2019, 23:20:52) [GCC 8.2.1 20181127]

**Comment by wutzkachel** [comment]
With aws cli it works!

aws ec2 run-instances --profile AWSCockpitMonitoringDevelEnv \
                      --image-id MyAmiID \
                      --subnet-id MySubnetID \
                      --security-group-ids MySgroupID \
                      --count 1 \
                      --instance-type t2.micro \
                      --key-name MyKey
